test/distributed/test_c10d_functional_native.py

```diff
@@ -222,6 +222,28 @@ def test_all_gather_into_tensor_single(self) -> None:
         assert output.eq(expect).all()
         assert output.completed
 
+    # https://github.com/pytorch/pytorch/issues/133421
+    @skip_if_lt_x_gpu(2)
+    def test_functional_collectives_inference_mode(self) -> None:
+        self._init_process_group()
+
+        with torch.inference_mode():
+            input = torch.full((2, 2), float(self.rank), device=self.device)
+            out1 = funcol.all_gather_tensor(
+                input, gather_dim=0, group=torch.distributed.group.WORLD
+            )
+            out2 = out1.to(dtype=torch.bfloat16)
+            # this tests that the call to .to() properly triggered a wait() on the AsyncCollectiveTensor
+            self.assertTrue(type(out2) is torch.Tensor)
+            self.assertEqual(
```

Comment:
@kwen2501 I wasn't able to repro the actual silent correctness in this test when I backed out the original fix (I confirmed that we we're calling `out1.to(...)` and doing a proper dtype conversion before issuing a proper `wait_tensor()` on the input, but the final output was still correct).

I did confirm that the earlier assert failed before though, which I think is a reasonable proxy - the output of calling `.to()` should give us a plain tensor, not an ACT, to represent the fact that we properly issued a sync.

Reply:
Hi @bdhirsh I think `assertEqual` would eventually trigger a wait on `out2` -- that's why you wouldn't see data corruption, with or without the fix.

In fact the return of ACT instead of Tensor is the only problem (copied from #152534):
> The issue seems to be that for an AsyncCollectiveTensor t, invoking t.float() does not trigger the wait_tensor, in which case it would return a regular torch.Tensor, but instead it returns a new AsyncCollectiveTensor with garbage data.

If the fix makes a difference on `self.assertTrue(type(out2) is torch.Tensor)`, then we are good.

Reply:
>  I think assertEqual would eventually trigger a wait on out2 -- that's why you wouldn't see data corruption, with or without the fix.

so we are agreed on this. What I'm surprised by, though, is that even if `assertEqual` issues a wait, that is still not early enough. If we issue the `.to()` dtype conversion kernel before running `wait_tensor`, I would have imagined that the `.to()` kernel would read garbage data (the allgather output buffer) before synchronizing.

Reply:
Either way, I agree that my `torch.Tensor` assertion is enough to convince myself that this test is ok

Reply:
But wait, wouldn't the transition from ACT to Tensor already trigger a wait? Are you saying that this is not observed in your test?